hypothesis-python/src/hypothesis/strategies/_internal/core.py

```diff
@@ -1082,8 +1082,18 @@ def as_strategy(strat_or_callable, thing, final=True):
         # The __optional_keys__ attribute may or may not be present, but if there's no
         # way to tell and we just have to assume that everything is required.
         # See https://github.com/python/cpython/pull/17214 for details.
-        optional = getattr(thing, "__optional_keys__", ())
-        anns = {k: from_type(v) for k, v in get_type_hints(thing).items()}
+        optional = set(getattr(thing, "__optional_keys__", ()))
+        anns = {}
+        for k, v in get_type_hints(thing).items():
+            if hasattr(v, "__origin__") and v.__origin__ in types.NotRequiredTypes:
+                optional.add(k)
+                anns.update({k: from_type(v.__args__[0])})
+            elif hasattr(v, "__origin__") and v.__origin__ in types.RequiredTypes:
+                anns.update({k: from_type(v.__args__[0])})
+            else:
+                anns.update({k: from_type(v)})
```

Comment:
This also made me realize that we're still missing a test case that a `Required` key on a `total=False` TypedDict *must* be included.

```suggestion
        anns = {}
        for k, v in get_type_hints(thing).items():
            origin = getattr(v, "__origin__", None)
            if origin in types.RequiredTypes + types.NotRequiredTypes:
                if origin in types.NotRequiredTypes:
                    optional.add(k)
                else:
                    optional.discard(k)
                try:
                    v = v.__args__[0]
                except IndexError:
                    raise InvalidArgument(f"`{k}: {v.__name__}` is not a valid type annotation") from None
            anns[k] = from_type(v)
```

Reply:
sure I can add that